tests/fs-tree-test.js

```diff
@@ -847,4 +850,107 @@ describe('FSTree', function() {
       });
     });
   });
+
+  describe('.applyPatch', function() {
+    it('applies all types of operations', function() {
+      var inputDir = 'tmp/fixture/input';
+      var outputDir = 'tmp/fixture/output';
+
+      fs.mkdirpSync(inputDir);
+      fs.mkdirpSync(outputDir);
+
+      var firstTree = FSTree.fromEntries(walkSync.entries(inputDir));
+
+      var fooIndex = path.join(inputDir, 'foo/index.js');
+      var barIndex = path.join(inputDir, 'bar/index.js');
+      var barOutput = path.join(outputDir, 'bar/index.js')
+
+      fs.outputFileSync(fooIndex, 'foo'); // mkdir + create
+      fs.outputFileSync(barIndex, 'bar'); // mkdir + create
+
+      var secondTree = FSTree.fromEntries(walkSync.entries(inputDir));
+      var patch = firstTree.calculatePatch(secondTree);
+
+      FSTree.applyPatch(inputDir, outputDir, patch);
+      expect(walkSync(outputDir)).to.deep.equal([
+        'bar/',
+        'bar/index.js',
+        'foo/',
+        'foo/index.js'
+      ]);
+      expect(fs.readFileSync(barOutput, 'utf-8')).to.equal('bar');
+
+      fs.removeSync(path.dirname(fooIndex)); // unlink + rmdir
+      fs.outputFileSync(barIndex, 'boo'); // change
+
+      var thirdTree = FSTree.fromEntries(walkSync.entries(inputDir));
+      patch = secondTree.calculatePatch(thirdTree);
+
+      FSTree.applyPatch(inputDir, outputDir, patch);
+      expect(walkSync(outputDir)).to.deep.equal([
+        'bar/',
+        'bar/index.js'
+      ]);
+      expect(fs.readFileSync(barOutput, 'utf-8')).to.equal('boo');
+
+      fs.removeSync('tmp');
```

Comment:
if one of the assertions in the current test block fails, this cleanup wont run and we will leave the suite in a bad state. I think we may want this in `afterEach`.